infrastructure/src/main/java/org/wso2/testgrid/infrastructure/providers/aws/utils/AwsAMIMapper.java

```diff
@@ -0,0 +1,160 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.testgrid.infrastructure.providers.aws.utils;
+
+import com.amazonaws.auth.PropertiesFileCredentialsProvider;
+import com.amazonaws.services.ec2.AmazonEC2;
+import com.amazonaws.services.ec2.AmazonEC2ClientBuilder;
+import com.amazonaws.services.ec2.model.DescribeImagesRequest;
+import com.amazonaws.services.ec2.model.DescribeImagesResult;
+import com.amazonaws.services.ec2.model.Image;
+import com.amazonaws.services.ec2.model.Tag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.testgrid.common.exception.TestGridInfrastructureException;
+import org.wso2.testgrid.common.util.StringUtil;
+import org.wso2.testgrid.common.util.TestGridUtil;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * This class provides a mapper to find the matching AMI based on the infrastructure parameters of the test-plan.
+ */
+public class AwsAMIMapper {
+    private static final String AWS_REGION_NAME = "us-east-1";
+    private static final Logger logger = LoggerFactory.getLogger(AwsAMIMapper.class);
+
+    private final AmazonEC2 amazonEC2;
+
+    public AwsAMIMapper() throws IOException, TestGridInfrastructureException {
+        Path configFilePath = Paths.get(TestGridUtil.getTestGridHomePath(), "config.properties");
+        if (!configFilePath.toFile().exists()) {
+            throw new TestGridInfrastructureException(
+                    "Configuration file not found. Hence can not provide AWS credentials.");
+        }
+        amazonEC2 = AmazonEC2ClientBuilder.standard()
+                .withCredentials(new PropertiesFileCredentialsProvider(configFilePath.toString()))
+                .withRegion(AWS_REGION_NAME)
+                .build();
+    }
+
+    /**
+     * This method finds out the relevant AMI-id of the AMI which matches for the infra-parameters passed.
+     * @param infraParameters Infrastructure parameters (of the test-plan)
+     * @return AMI-id of the matching AMI
+     * @throws TestGridInfrastructureException When can not find a matching AMI
+     */
+    public String getAMI(Properties infraParameters) throws TestGridInfrastructureException {
+        List<Image> amiList = getAMIListFromAWS();
+        if (amiList.isEmpty()) {
+            throw new TestGridInfrastructureException("List of AMIs is empty. Hence can not find a matching AMI.");
+        }
+        //AMIs are containing tags only for the lookup-parameters.
+        Properties lookupParameters = filterLookupParameters(infraParameters);
+
+        Boolean isMissingLookupParams;  //False means one or more lookup parameters are missing.
+        String amiId = null;
+
+        //Check for an AMI which contains all (tags) lookupParameters.
```

Comment:
If you need to add a in-line comment, then that means you probably need to extract that logic into a private method.

Usually, if the code block is more than 4 lines, you should try to move it to a separate method. (yes, 4! :) )

Reply:
Some refs: https://softwareengineering.stackexchange.com/a/195992/33655